### ISSUE TITLE
Truncated answers and DNS over TCP

### ISSUE THREAD

**Issue by quaqo** [IngoZenz/personaldnsfilter#161]
Hello there. First and foremost: thanks for your great work. I really enjoy how simple and efficient PersonalDNSFilter is.
I also appreciated the fact that I can tweak it on Android with a configuration file!

Anyways, hoping to be of help: I had a problem on Android and managed to debug it on my desktop.

I was trying to get the IP address for _app.powerbi.com_ (which is a long list of nested CNAMEs) and it simply didn't work with my provider's DNS server. I thought of a issue with CNAME cloaking, and verified with the source code that it is not.

On my computer it worked with many of the fallback DNS servers and I managed to pinpoint the exact cause using Wireshark.

It seems that many servers (e.g. 94.140.14.140 or 8.8.8.8) always return a "compact" answer, so that if the connection is started on UDP, the answer is always under the 512 byte limit of DNS over UDP.

My provider's DNS server instead (and many others) provide a truncated answer @ 512 bytes with the TC flag set, so that the client knows to reconnect with TCP to get the full message.

What happens with PersonalDNSFilter is that the software correctly rely 1-1 the DNS answer, but then something's wrong with the TCP connection and it fails.

I tried explicitly putting: fallbackDNS = [X.X.X.X]::53::TCP in the configuration file and indeed everything worked as intended.

For you to replicate the "bug" I found a public DNS that seems to truncate: 185.30.211.154.

So, setting:

`fallbackDNS = 185.30.211.154`

I get:

```
$ nslookup -port=53 app.powerbi.com 127.0.0.1
;; Truncated, retrying in TCP mode.
;; Connection to 127.0.0.1#53(127.0.0.1) for app.powerbi.com failed: connection refused.
```

For example, with Google's:

```
$ nslookup app.powerbi.com 8.8.8.8
Server:        8.8.8.8
Address:    8.8.8.8#53

Non-authoritative answer:
app.powerbi.com    canonical name = app.privatelink.analysis.windows.net.
app.privatelink.analysis.windows.net    canonical name = 997de1ee-c405-4364-8b90-eb6f601a6af2.trafficmanager.net.
997de1ee-c405-4364-8b90-eb6f601a6af2.trafficmanager.net    canonical name = pbi-wfe-switzerland-north.powerbi.com.
pbi-wfe-switzerland-north.powerbi.com    canonical name = pbi-wfe-switzerland-north-ase.powerbi.com.
pbi-wfe-switzerland-north-ase.powerbi.com    canonical name = app-pbi-wfe-switzerland-north.pbi-wfe-switzerland-north-ase.p.azurewebsites.net.
app-pbi-wfe-switzerland-north.pbi-wfe-switzerland-north-ase.p.azurewebsites.netcanonical name = waws-prod-zrh-c2dc13a7.sip.p.azurewebsites.windows.net.
waws-prod-zrh-c2dc13a7.sip.p.azurewebsites.windows.net    canonical name = waws-prod-zrh-c2dc13a7.cloudapp.net.
Name:    waws-prod-zrh-c2dc13a7.cloudapp.net
Address: 51.103.139.236
```

Could it be looked into? 

Let me know if I can help in any ways. I don't have a java environment available so I didn't start playing with the source code.

Thank you.

**Comment by IngoZenz** [comment]
thanks a lot for this detailed analysis and the hint for reproducing it. I was aware already of the limitation which rarely causes issues..
Will think about on how to handle it. I would like to avoid to handle TCP packages over the VPN interface...

**Comment by quaqo** [comment]
Thanks for the fast reply. I gave a quick look to the resolver source code and it's my understanding that with (at least) the local resolver on, you have no issue whatsoever on getting the actual IP from the original server. The problem lies between PersonalDNSFilter on the VPN, where you only listen on UDP.

This kind of issue is present with many Microsoft domains (e.g. Office365), because of the long list of CNAMEs.

It seems that 8.8.8.8 replies with [name compression ](https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.4) on and that can keep some of the datagrams below the 512 size.

Honestly I don't know if it would be better to not relay the message directly and instead build a new response from scratch (using name compression) or just listening on TCP (which I agree has it's own problems).

I don't see many options to tackle this.

I encountered the error only because of COVID a state app is relying on Microsoft Power BI and I could SEE that wasn't working, but after trying to debug it I understood why other apps were not working as intended and I had to whitelist them (which obviously I wouldn't want to do...).

Thanks again!

P.S.: I love that you give the option to route selected IPs to the local server over the VPN.

**Comment by IngoZenz** [comment]
Thanks again. So it worked for you when you used a DOT or DOH DNS as fallback DNS server within pDNSf? If so, seems the UDP DNS client  does accept longer DNS response packages via UDP. 

I also thought already about the option to return an own generated response with reduced size...

**Comment by quaqo** [comment]
Not exactly. My thought was that If I had used another server via DoT it wouldn't have been a "significant" experiment.

What I did is force pDNSf in TCP mode to the very same server on port 53.

In this way the server sent the whole message without truncating it, but yes you're correct: the relayed message by pDNSf was on UDP and was accepted.

But this isn't standard compliant unless used with EDNS0 I think so YMMV.

**Comment by IngoZenz** [comment]
I know it  snot standard but working for most clients ...could be pragmatic workaround for now ...
In addition we could have the compressed response generated by pDNSf ... 
However I would need time - which I do not really have currently ;-)

**Comment by quaqo** [comment]
I tried something more and this is what happens:

**pDNSf set with nameserver in UDP**

```
nslookup [127.0.0.1]    -> pDNSf [127.0.0.1]    UDP        query A app.powerbi.com
pDNSf            -> upstreamDNS        UDP        query A app.powerbi.com
upstreamDNS        -> pDNSf        UDP    554b    incomplete query response (TC bit set because 554b > 512b)
pDNSf [127.0.0.1]    -> nslookup [127.0.0.1]    UDP    554b  incomplete query response (TC bit set, relayed as it is)
nslookup [127.0.0.1]    -> pDNSf [127.0.0.1]    TCP        SYN (reads TC bit, tries to get the rest of the response via TCP, as per DNS RFC)
pDNSf [127.0.0.1]    -> nslookup [127.0.0.1]    TCP        RST
```


**pDNSf set with nameserver in TCP**

```
nslookup [127.0.0.1]    -> pDNSf [127.0.0.1]    UDP        query A app.powerbi.com
pDNSf            -> upstreamDNS        TCP        query A app.powerbi.com
upstreamDNS        -> pDNSf        TCP    797b    query response (TRUNCATED flag set to 0, because we're in TCP)
pDNSf [127.0.0.1]    -> nslookup [127.0.0.1]    UDP    797b  query response (TRUNCATED flag set to 0, we're in UDP with 797b > 512b, but nslookup doesn't seem to care as long as it receives a complete response)
```

I also tried querying 8.8.8.8: the answer is below 512b because it drops the "Authoritative nameservers" part.

It seems confirmed that every client expects TCP, as per RFC1123 sec 6.1.3.2 as updated by RFC776, "All general-purpose DNS implementations MUST support both UDP and TCP transport."

The clients I tried (on Linux and Android) seems to not care if they receive bigger packets (as long as the underlying system allows routing of bigger UPD datagrams).

They must receive the full answer via UDP without the TC flag and they accept it, even thou this is a non-standard implementation of DNS over UDP.

This is also what happens right now with DoT or DoH: you may send DNS over UDP packets over the LAN that exceed the standard 512b limit.


What I suggest would be to not limit pDNSf to an UDP only resolver on upstream, instead it should behave like a standard client to get the full response, then a "hack" could be to continue serving a non standard answer over the VPN via UDP (as it seems to be accepted).

What I mean is:

1) pDNSf receives the query
2) pDNSf relays the query
3) pDNSf receives the answer with TrunCated bit set to 1
4) pDNSf queries again the upstream server in TCP mode (instead of UDP)
5) pDNSf receives the answer via TCP
6) pDNSf serves the non standard answer via UDP over the VPN


What do you think?

**Comment by quaqo** [comment]
I'd add that although I'm not very well versed in Java, if you could point me in the direction of how to have an environment to build the source code on Linux, I could try to write a patch and propose it to you.

**Comment by IngoZenz** [comment]
Yes! Your proposal makes sense - was thinking the same...
In order to build I suggest to install the Android Studio ...

**Comment by IngoZenz** [comment]
Guess it should be implemented in https://github.com/IngoZenz/personaldnsfilter/blob/master/app/src/main/java/dnsfilter/DNSServer.java within the UDP implementation.

**Comment by quaqo** [comment]
Yep. I was looking at it. The code is very well commented, so it was easy to get there. Thank you! I'll report back as soon as I have time to "play" with it. :-)
